src/Radicle/Internal/Core.hs

```diff
@@ -375,11 +441,77 @@ instance GhcExts.IsList (PrimFns m) where
-    , bindingsPrimFns :: prims
-    , bindingsRefs    :: IntMap Value
-    , bindingsNextRef :: Int
-    } deriving (Eq, Show, Functor, Generic)
+    { bindingsEnv            :: Env Value
+    , bindingsPrimFns        :: prims
+    , bindingsRefs           :: IntMap Value
+    , bindingsNextRef        :: Int
+    , bindingsHandles        :: IntMap Handle.Handle
+    , bindingsNextHandle     :: Int
+    , bindingsProcHandles    :: IntMap ProcessHandle
+    , bindingsNextProcHandle :: Int
+    } deriving (Functor, Generic)
+
+emptyBindings :: Bindings (PrimFns m)
+emptyBindings = Bindings mempty mempty mempty 0 mempty 0 mempty 0
+
+-- | Extract an environment and references from a Radicle value and put
+-- them as the current bindings. Primitive functions are not changed.
+--
+-- Throws 'OtherError' if @value@ cannot be parsed.
+--
+-- prop> gets bindingsToRadcile >>= setBindings = pure ()
+setBindings :: Monad m => Value -> Lang m ()
+setBindings value = do
+    case bindingsFromRadicle value of
+        Left e  -> throwErrorHere $ OtherError $ "Invalid bindings: " <> e
+        Right newBindings -> do
+            prims <- gets bindingsPrimFns
+            handles <- gets bindingsHandles
+            nextHandle <- gets bindingsNextHandle
+            procHandles <- gets bindingsProcHandles
+            nextProcHandle <- gets bindingsNextProcHandle
+            put $ newBindings
+                { bindingsPrimFns = prims
+                , bindingsHandles = handles
+                , bindingsNextHandle = nextHandle
+                , bindingsProcHandles = procHandles
+                , bindingsNextProcHandle = nextProcHandle
+                }
+  where
+    bindingsFromRadicle x = case x of
+        Dict d -> do
+            env' <- kwLookup "env" d ?? "Expecting 'env' key"
+            refs' <- kwLookup "refs" d ?? "Expecting 'refs' key"
+            refs <- makeRefs refs'
+            env <- envFromRad env'
+            pure $ emptyBindings
+                { bindingsEnv = env
+                , bindingsRefs = refs
+                , bindingsNextRef = length refs
+                }
+        _ -> throwError "Expecting dict"
+    makeRefs refs = case refs of
+        List ls -> pure (IntMap.fromList $ zip [0..] ls)
+        _       -> throwError $ "Expecting dict"
+    envFromRad env = case env of
+        Dict d -> fmap (Env . Map.fromList)
+                $ forM (Map.toList d) $ \(k, v) -> case k of
+            Atom i -> (i, ) <$> fromRad v
+            k'     -> Left $ "Expecting atom keys. Got: " <> show k'
+        _ -> Left "Expecting dict"
+
+-- | Serializes the environment and references into a Radicle value.
+--
+-- prop> gets bindingsToRadcile >>= setBindings = pure ()
+bindingsToRadicle :: Bindings a -> Value
+bindingsToRadicle x =
+    Dict $ Map.fromList
+        [ (Keyword $ Ident "env", env)
+        , (Keyword $ Ident "refs", refs)
+        ]
+  where
+    env = Dict . Map.mapKeys Atom . Map.map toRad . fromEnv $ bindingsEnv x
```

Comment:
Map.mapWithKey

Reply:
Hm, not sure I get it. We're mapping the keys and the values, but mapWithKey takes the key and the value as arg but only changes the value

Reply:
Indeed ignore me